### PR TITLE
Throw exception if rcl_timer_init fails

### PR DESCRIPTION
**Pull request by brawner**
This is the second PR related to https://github.com/ros2/rclcpp/issues/1177. I believe along with #1178 the issue will be sufficiently addressed. 

Because the initialized `rcl_timer_t` is invalid, no other function calls in TimerBase will be successful. the destructor of `GenericTimer` in fact calls cancel, which will fail. Therefore, I think it's best to just throw the exception up front in the constructor of TimerBase.

The test conflicts with the test in #1178 and whichever is merged second will need to be updated beforehand.

### CODE REVIEW

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11194/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6472/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9156/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11113/)

File: rclcpp/test/rclcpp/test_create_timer.cpp [hunk 1]
```diff
@@ -61,3 +61,40 @@ TEST(TestCreateTimer, call_with_node_wrapper_compiles)
     []() {});
   rclcpp::shutdown();
 }
+
+TEST(TestCreateTimer, call_wall_timer_with_bad_arguments)
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_create_timer.cpp:65]: @brawner consider adding a test that forces a throw on `Timer` construction using a bad clock.

Comment (brawner) [reply rclcpp/test/rclcpp/test_create_timer.cpp:65]: Unfortunately, I think that's harder than it sounds. The constructor of WallTimer uses RCL_STEADY_CLOCK, but GenericTimer can use an arbitrary clock type. However, the constructor of Clock calls `rcl_clock_init`, which will validate it's one argument, `clock_type`. So I'm not sure it's feasible to pass in a bad clock without mocking the rclcpp::Clock class.

It's possible I'm missing something, do you have any suggestions along these lines?

Comment (hidmic) [reply rclcpp/test/rclcpp/test_create_timer.cpp:65]: I think [`RCL_CLOCK_UNINITIALIZED`](https://github.com/ros2/rcl/blob/f84592fb2f6c93cb25dd4ca59b16858cf0d9abdf/rcl/include/rcl/time.h#L61) may serve your purposes.

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11204/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6478/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9164/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11122/)

File: rclcpp/test/rclcpp/test_create_timer.cpp (hunk 1 above)
Comment (brawner) [reply rclcpp/test/rclcpp/test_create_timer.cpp:65]: Added this test to test_timer.cpp

**Comment by brawner** [comment]
Rebased on to master after merging #1178
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11212/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6484/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9172/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11129/)

File: rclcpp/test/rclcpp/test_timer.cpp [hunk 2]
```diff
@@ -151,3 +152,47 @@ TEST_F(TestTimer, test_run_cancel_timer)
...
+  auto callback = []() {};
+  auto node_base = rclcpp::node_interfaces::get_node_base_interface(test_node);
+  auto context = node_base->get_context();
+
+  auto steady_clock = std::make_shared<rclcpp::Clock>(RCL_STEADY_TIME);
+
+  // Negative period
+  EXPECT_THROW(
+    rclcpp::GenericTimer<decltype(callback)>(
+      steady_clock, -1ms, std::forward<decltype(callback)>(callback), context),
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_timer.cpp:166]: @brawner why forwarding in this context? BTW the [use of `std::move` in `WallTimer` constructor on a `FunctorT&&` argument](https://github.com/ros2/rclcpp/blob/4f14a0cc142f43c62e40e5fadca2528183fc0a48/rclcpp/include/rclcpp/timer.hpp#L275) seems wrong and there I would've used forwarding.

Comment (brawner) [reply rclcpp/test/rclcpp/test_timer.cpp]: Switched std::forward to std::move, here and below.

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11237/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6503/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9198/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11151/)

File: rclcpp/test/rclcpp/test_timer.cpp [hunk 3]
```diff
@@ -151,3 +152,43 @@ TEST_F(TestTimer, test_run_cancel_timer)
...
+TEST_F(TestTimer, test_bad_arguments) {
+  auto callback = []() {};
+  auto node_base = rclcpp::node_interfaces::get_node_base_interface(test_node);
+  auto context = node_base->get_context();
+
+  auto steady_clock = std::make_shared<rclcpp::Clock>(RCL_STEADY_TIME);
+
+  // Negative period
+  EXPECT_THROW(
+    rclcpp::GenericTimer<decltype(callback)>(steady_clock, -1ms, std::move(callback), context),
```
Comment (hidmic) [inline rclcpp/test/rclcpp/test_timer.cpp:165]: @brawner here and elsewhere, no need for `std::move`:
```suggestion
    rclcpp::GenericTimer<decltype(callback)>(steady_clock, -1ms, callback, context),
```

Sorry if it seemed like I was suggesting so.

Comment (brawner) [reply rclcpp/test/rclcpp/test_timer.cpp]: An rvalue reference is required to pass callback and I was using the decltype() to be lazy about the template instantiation. But I'm not sure what the expected behavior is of passing std::move on an empty callback multiple times. Switching template type to actual type, and passing in an empty callback.

**Comment by brawner** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11241/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6505/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9201/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11153/)

**Pull request merged**